File: C&Python/Samsung/보호 필름.py
min_value = 210000000
T = int(input())
B = [[0] * 14 for i in range(21)]
D = 0
K = 0
W = 0


def TEST(BOARD):
    check = [0] * W
    for i in range(W):
        cnt = 1
        for j in range(1, D, 1):
            if (BOARD[j - 1][i] == BOARD[j][i]):
                cnt += 1
                if (cnt >= K):
                    check[i] = 1
                    break
            else:
                cnt = 1
        if (check[i] == 0):
            return False
    return True


def DFS(BOARD, check, cnt, row):
    global min_value
    if min_value != 210000000 and min_value < cnt:
        return

    if (TEST(BOARD) == True):
        #print(BOARD)
        if min_value > cnt:
            min_value = cnt
            return
    
    # exit condition
    if (row >= D) or (cnt >= D):
        return

    # not A, B
    DFS(BOARD,check, cnt, row + 1)

    # A
    if check[row] == 0:
        backup = [0] * 21
        check[row] = 1
        for i in range(W):
            backup[i] = BOARD[row][i]
            BOARD[row][i] = 0
        DFS(BOARD, check, cnt + 1, row + 1)
        for i in range(W):
            BOARD[row][i] = backup[i]
        check[row] = 0

    # B
    if check[row] == 0:
        backup = [0] * 21
        check[row] = 1
        for i in range(W):
            backup[i] = BOARD[row][i]
            BOARD[row][i] = 1
        DFS(BOARD, check, cnt + 1, row + 1)
        for i in range(W):
            BOARD[row][i] = backup[i]
        check[row] = 0


for test_case in range(1, T + 1):
    check = [0] * 21
    AA, BB, CC = map(int,input().split())
    D = AA
    W = BB
    K = CC
    B = [list(map(int, input().split())) for _ in range(D)]
    if K == 1:
        print("#" + str(test_case) + " 0")
    else:
        DFS(B, check, 0, 0)
        print("#" + str(test_case) + " " + str(min_value))
        min_value = 210000000

File: C++/Baekjoon/Samsung/보호필름.cpp
#include <iostream>
#include <cstdio>
using namespace std;
int D, W, K;
int check[14]; // A or B 약품 사용여부
pair<int, int> answer = { 1000000000,1000000000 };
bool Test(int Board[14][21]);
void DFS(int v[14][21], int acnt, int bcnt, int row);

int main(int argc, char** argv)
{
	ios_base::sync_with_stdio(NULL);
	cin.tie(NULL);
	cout.tie(NULL);
	int test_case;
	int T;
	cin >> T;

	for (test_case = 1; test_case <= T; ++test_case)
	{
		int Board[14][21];
		cin >> D >> W >> K;
		for (int i = 0; i < D; i++)
		{
			for (int j = 0; j < W; j++)
			{
				cin >> Board[i][j];
			
			}
		}
		DFS(Board, 0, 0, 0);
		cout << "#" << test_case << " " << answer.first + answer.second << '\n';

		answer.first = 1000000000;
		answer.second = 1000000000;

	}
	return 0;
}


bool Test(int Board[14][21])
{
	int test_check[21];
	for (int i = 0; i < 21; i++)
		test_check[i] = 0;

	for (int i = 0; i < W; i++)
	{
		int cnt = 1;
		for (int j = 1; j < D; j++)
		{
			if (Board[j - 1][i] == Board[j][i])
			{
				cnt++;
				if (cnt >= K)
				{
					test_check[i] = 1;
					break;
				}
			}
			else
			{
				if (cnt >= K)
				{
					test_check[i] = 1;
					break;
				}
				cnt = 1;
			}
		}

		if (test_check[i] == 0)
			return false;
	}
	for (int i = 0; i < W; i++)
	{
		if (test_check[i] == 0)
			return false;
	}
	return true;
}

void DFS(int v[14][21], int acnt, int bcnt, int row)
{
	/* exit condition */

	if (answer.first != 1000000000 && answer.first + (answer.first + answer.second) < (acnt + bcnt))
		return;

	if (Test(v) == true)
	{
		if ((answer.first + answer.second) > (acnt + bcnt))
		{
			answer.first = acnt;
			answer.second = bcnt;
		}
	}

	if (row >= D)
		return;

	if ((acnt + bcnt) >= D)
	{
		return;
	}

	/* not A, B */
	DFS(v, acnt, bcnt, row + 1);

	/* A */
	if (check[row] == 0)
	{
		int backup[21];
		check[row] = 1;
		for (int i = 0; i < W; i++)
		{
			backup[i] = v[row][i];
			v[row][i] = 0;
		}

		DFS(v, acnt + 1, bcnt, row + 1);

		check[row] = 0;
		for (int i = 0; i < W; i++)
		{
			v[row][i] = backup[i];
		}
	}

	/* B */
	if (check[row] == 0)
	{
		int backup[21];
		check[row] = 1;
		for (int i = 0; i < W; i++)
		{
			backup[i] = v[row][i];
			v[row][i] = 1;
		}

		DFS(v, acnt, bcnt + 1, row + 1);
		check[row] = 0;
		for (int i = 0; i < W; i++)
		{
			v[row][i] = backup[i];
		}
	}
}